src/shims/x86/sse42.rs

```diff
@@ -0,0 +1,480 @@
+            // https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ig_expand=1046,1047
+            "pcmpestriz128" | "pcmpestris128" => {
+                let [_, len1, _, len2, imm] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                let len = if unprefixed_name == "pcmpestris128" { len1 } else { len2 };
+                let len = this.read_scalar(len)?.to_i32()?;
+                let imm = this.read_scalar(imm)?.to_u8()?;
+                this.write_scalar(
+                    Scalar::from_i32(i32::from(len < default_len::<i32>(imm))),
+                    dest,
+                )?;
+            }
+
+            // Used to implement the `_mm_crc32_u{8, 16, 32, 64}` functions.
+            // These functions calculate a 32-bit CRC using `0x11EDC6F41`
+            // as the polynomial, also known as CRC32C.
+            // https://datatracker.ietf.org/doc/html/rfc3720#section-12.1
+            "crc32.32.8" | "crc32.32.16" | "crc32.32.32" | "crc32.64.64" => {
+                let bit_size = match unprefixed_name {
+                    "crc32.32.8" => 8,
+                    "crc32.32.16" => 16,
+                    "crc32.32.32" => 32,
+                    "crc32.64.64" => 64,
+                    _ => unreachable!(),
+                };
+
+                if bit_size == 64 && this.tcx.sess.target.arch != "x86_64" {
+                    return Ok(EmulateItemResult::NotSupported);
+                }
+
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                let left = this.read_scalar(left)?;
+                let right = this.read_scalar(right)?;
+
+                let crc = if bit_size == 64 {
+                    // The 64-bit version will only consider the lower 32 bits,
+                    // while the upper 32 bits get discarded.
+                    #[allow(clippy::cast_possible_truncation)]
+                    u128::from((left.to_u64()? as u32).reverse_bits())
+                } else {
+                    u128::from(left.to_u32()?.reverse_bits())
+                };
+                let v = match bit_size {
+                    8 => u128::from(right.to_u8()?.reverse_bits()),
+                    16 => u128::from(right.to_u16()?.reverse_bits()),
+                    32 => u128::from(right.to_u32()?.reverse_bits()),
+                    64 => u128::from(right.to_u64()?.reverse_bits()),
+                    _ => unreachable!(),
+                };
+
+                // Perform polynomial division modulo 2.
+                let mut dividend = (crc << bit_size) ^ (v << 32);
+                const POLYNOMIAL: u128 = 0x11EDC6F41;
+                let leading_zeros = dividend.leading_zeros();
+                while leading_zeros <= POLYNOMIAL.leading_zeros() {
+                    dividend ^= (POLYNOMIAL << POLYNOMIAL.leading_zeros()) >> leading_zeros;
+                }
```

Comment:
This looks like dark magic to me.^^ Shifting by the number of leading zeroes of `(crc << bit_size) ^ (v << 32)`... now my head hurts. ;)

I assume this is a standard algorithm; is there a nice website we can link to that documents/explains these shifts?

Reply:
The intrinsics guide goes into the previous steps and mentions them. _Why_ they are done I don't know either, for that my grasp on CRC is too weak.

As for the loop, its simple schoolbook division, but it's in GF(2), so addition/subtraction is done using a bitwise XOR.

Reply:
The intrinsic guide being [this](https://datatracker.ietf.org/doc/html/rfc3720#section-12.1)? It doesn't seem to say anything specific about this intrinsic though, just CRC in general.

> As for the loop, its simple schoolbook division, but it's in GF(2), so addition/subtraction is done using a bitwise XOR.

I believe that it is standard, but having a link to somewhere that explains the concrete algorithm, and in particular the loop condition involving `leading_zeros`, would still be good.

Reply:
I didn't find any satisfactory explanation online so I hope the my custom explanation is sufficient.

I also decided to extract the algorithm and exhaustively compare it with the machine results for the `_mm_crc32_u8` intrinsic. The algorithm is _definitely_ correct.

Reply:
Oh, I thought you had copied the algorithm from somewhere and it'd be easier to link to that. I didn't realize you just implemented it based on the abstract algebraic spec yourself. :) Thanks for adding the comment!